plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
+		return nil, errors.New("core values cannot be duplicated")
 	}
+	return parsedCores, nil
+}
 
-	if !p.skipFirstIteration {
-		if p.cpuC0StateResidency || p.cpuBusyCycles {
-			p.addCPUC0StateResidencyMetric(cpuID, acc)
-		}
-
-		if p.cpuC1StateResidency {
-			p.addCPUC1StateResidencyMetric(cpuID, acc)
-		}
+// parseGroupCores takes a string which represents a group of one or more
+// CPU IDs (e.g. "0", "1-3", or "4,5,6") and returns a slice of integers with
+// all CPU IDs within the group.
+func parseGroupCores(coreGroup string) ([]int, error) {
+	coreElems := strings.Split(coreGroup, ",")
+	cores := make([]int, 0, len(coreElems))
 
-		if p.cpuC6StateResidency {
-			p.addCPUC6StateResidencyMetric(cpuID, acc)
+	for _, coreElem := range coreElems {
+		if strings.Contains(coreElem, "-") {
+			pCores, err := parseCoreRange(coreElem)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse core range %q: %w", coreElem, err)
+			}
+			cores = append(cores, pCores...)
+		} else {
+			singleCore, err := strconv.Atoi(coreElem)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse single core: %w", err)
+			}
+			cores = append(cores, singleCore)
 		}
 	}
+	return cores, nil
 }
 
-func (p *PowerStat) addCPUFrequencyMetric(cpuID string, acc telegraf.Accumulator) {
-	frequency, err := p.msr.retrieveCPUFrequencyForCore(cpuID)
+// parseCoreRange takes a string representing a core range (e.g. "0-4"), and
+// returns a slice of integers with all elements within this range.
+func parseCoreRange(coreRange string) ([]int, error) {
+	rangeVals := strings.Split(coreRange, "-")
+	if len(rangeVals) != 2 {
+		return nil, errors.New("invalid core range format")
+	}
 
-	// In case of an error leave func
+	low, err := strconv.Atoi(rangeVals[0])
 	if err != nil {
-		p.Log.Debugf("Error while reading file: %v", err)
-		return
+		return nil, fmt.Errorf("failed to parse low bounds' core range: %w", err)
 	}
 
-	cpu := p.cpuInfo[cpuID]
-	tags := map[string]string{
-		"package_id": cpu.physicalID,
-		"core_id":    cpu.coreID,
-		"cpu_id":     cpu.cpuID,
+	high, err := strconv.Atoi(rangeVals[1])
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse high bounds' core range: %w", err)
 	}
 
-	fields := map[string]interface{}{
-		"cpu_frequency_mhz": roundFloatToNearestTwoDecimalPlaces(frequency),
+	if high < low {
+		return nil, errors.New("high bound of core range cannot be less than low bound")
 	}
-
-	acc.AddGauge("powerstat_core", fields, tags)
+	return makeCores(low, high), nil
 }
 
-func (p *PowerStat) addCPUTemperatureMetric(cpuID string, acc telegraf.Accumulator) {
-	coresData := p.msr.getCPUCoresData()
-	temp := coresData[cpuID].throttleTemp - coresData[cpuID].temp
+// makeCores takes a low and high bounds of an interval of integers, and returns
+// a slice will all integers within the interval.
+func makeCores(low, high int) []int {
+	cores := make([]int, high-low+1)
+	for i := range cores {
+		cores[i] = i + low
+	}
+	return cores
+}
```

Comment:
Please merge the two!

Reply:
Makes sense. Done as part of commit [e2de6f6b](https://github.com/influxdata/telegraf/pull/14363/commits/e2de6f6b2577fe4e9ec7b420132b2750fe7e2654).